pyopencl/invoker.py

```diff
@@ -315,6 +315,12 @@ def generate_enqueue_and_set_args(function_name,
         arg_types,
         work_around_arg_count_bug, warn_about_arg_count_bug):
 
+    if num_cl_args > 500:
```

Comment:
[CL_​DEVICE_​MAX_​PARAMETER_​SIZE](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#CL_DEVICE_MAX_PARAMETER_SIZE)

Reply:
https://github.com/pocl/pocl/pull/1046 ;-)

Is it ok to estimate the argument size based on `num_cl_args*8` or should I look at each argument's actual size (how can I do that for a VectorArg?)?

Reply:
This is not a performance-sensitive bit of code, so I see no reason to not at least try to be precise.

>  (how can I do that for a VectorArg?)?

https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_ADDRESS_BITS

Reply:
An issue I couldn't resolve: the invoker is device-independent, but [CL_​DEVICE_​MAX_​PARAMETER_​SIZE](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#CL_DEVICE_MAX_PARAMETER_SIZE) is not. How can we resolve this?

- Somehow pass the device to `generate_enqueue_and_set_args` (how?)
- `gen()` the argument checking code into the invoker (potential performance hit, also probably too late to meaningfully warn about the arg count)
- Add the argument checking code somewhere else (where?)

Reply:
A kernel is associated with a context, and from a context you can find the devices.

Reply:
Thanks! What do you think of 24d00a40250033f1ffedf09fbc6d727d1dd13c1b?